src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotHideBaseClassMethods.cs

```diff
@@ -0,0 +1,115 @@
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftApiDesignGuidelinesAnalyzersResources.DoNotHideBaseClassMethodsDescription), MicrosoftApiDesignGuidelinesAnalyzersResources.ResourceManager, typeof(MicrosoftApiDesignGuidelinesAnalyzersResources));
+
+        public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Design,
+            DiagnosticSeverity.Warning, 
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            helpLinkUri: "https://msdn.microsoft.com/en-us/library/ms182143.aspx",
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.RegisterSymbolAction(SymbolAnalyzer, SymbolKind.Method);
+        }
+
+        private void SymbolAnalyzer(SymbolAnalysisContext context)
+        {
+            var method = (IMethodSymbol)context.Symbol;
+
+            // Bail out if this method overrides another (parameter types must match) or doesn't have any parameters
+            if (method.IsOverride || !method.Parameters.Any())
+            {
+                return;
+            }
+
+            foreach (var hiddenMethod in GetMethodsHiddenByMethod(method, method.ContainingType.BaseType))
+            {
+                var diagnostic = Diagnostic.Create(Rule, context.Symbol.Locations[0], method.ToDisplayString(), hiddenMethod.ToDisplayString());
+                context.ReportDiagnostic(diagnostic);
+            }
+        }
+
+        private IEnumerable<IMethodSymbol> GetMethodsHiddenByMethod(IMethodSymbol method, INamedTypeSymbol baseType)
+        {
+            while (true)
+            {
+                if (baseType?.BaseType == null)
+                {
+                    // We must be at System.Object, and the methods in System.Object cannot be hidden
+                    yield break;
+                }
+
+                var baseMethods = baseType.GetMembers(method.Name)
+                    .OfType<IMethodSymbol>()
+                    .Where(x => !(x.IsStatic || x.IsVirtual)
+                           && x.ReturnType.Equals(method.ReturnType)
+                           && x.DeclaredAccessibility != Accessibility.Private
+                           && x.Parameters.Length == method.Parameters.Length);
+
+                foreach (var baseMethod in baseMethods)
+                {
+                    var isMethodHidden = false;
+
+                    for (var i = 0; i < baseMethod.Parameters.Length; ++i)
```

Comment:
There is still a subtle bug here: if first param types are unrelated but last ones have base-derived relationship, you will return true instead of false.